File: typescript/systemdesign/trieAutocomplete.ts
/*
Trie-based prefix autocomplete: inserts words into a trie, tracks a
frequency/weight per word, and returns the top-K matches for a given prefix
ranked by weight then lexicographically.

status - completed
*/

import { logp } from "../utils/logger";

class TrieNode {
  readonly children: Map<string, TrieNode> = new Map();
  isWord = false;
  weight = 0;
}

class TrieAutocomplete {
  private readonly root = new TrieNode();

  insert(word: string, weight: number = 1): void {
    let node = this.root;
    for (const ch of word) {
      let next = node.children.get(ch);
      if (!next) {
        next = new TrieNode();
        node.children.set(ch, next);
      }
      node = next;
    }
    node.isWord = true;
    node.weight += weight;
  }

  private collectWords(node: TrieNode, prefix: string, out: Array<{ word: string; weight: number }>): void {
    if (node.isWord) {
      out.push({ word: prefix, weight: node.weight });
    }
    for (const [ch, child] of node.children) {
      this.collectWords(child, prefix + ch, out);
    }
  }

  topMatches(prefix: string, limit: number = 5): string[] {
    let node = this.root;
    for (const ch of prefix) {
      const next = node.children.get(ch);
      if (!next) {
        return [];
      }
      node = next;
    }

    const matches: Array<{ word: string; weight: number }> = [];
    this.collectWords(node, prefix, matches);

    matches.sort((a, b) => b.weight - a.weight || a.word.localeCompare(b.word));
    return matches.slice(0, limit).map((m) => m.word);
  }
}

// usage scenarios

const trie = new TrieAutocomplete();
const seed: Array<[string, number]> = [
  ["car", 10],
  ["care", 5],
  ["career", 8],
  ["cart", 3],
  ["cared", 2],
  ["dog", 4],
];
for (const [word, weight] of seed) {
  trie.insert(word, weight);
}

logp(`autocomplete("car") -> ${JSON.stringify(trie.topMatches("car", 3))}`);
logp(`autocomplete("care") -> ${JSON.stringify(trie.topMatches("care", 3))}`);
logp(`autocomplete("d") -> ${JSON.stringify(trie.topMatches("d"))}`);
logp(`autocomplete("zzz") -> ${JSON.stringify(trie.topMatches("zzz"))} (expected empty, no such prefix)`);

File: python/systemdesign/trie_autocomplete.py
"""
Trie-based prefix autocomplete. Words are inserted into a trie with
frequency counts at terminal nodes, and lookups walk to the prefix
node then collect the top matches ranked by frequency then
alphabetically.

status - completed
"""

from base_logger.logging_event import create_logger


class TrieNode:

    def __init__(self):
        self.children = {}
        self.is_word = False
        self.frequency = 0


class TrieAutocomplete:

    log = create_logger(__name__)

    def __init__(self):
        self.root = TrieNode()

    def insert(self, word: str, frequency: int = 1):
        node = self.root
        for ch in word:
            node = node.children.setdefault(ch, TrieNode())
        node.is_word = True
        node.frequency += frequency

    def _collect_words(self, node, prefix, results):
        if node.is_word:
            results.append((prefix, node.frequency))
        for ch, child in node.children.items():
            self._collect_words(child, prefix + ch, results)

    def top_matches(self, prefix: str, k: int = 5):
        node = self.root
        for ch in prefix:
            if ch not in node.children:
                return []
            node = node.children[ch]

        results = []
        self._collect_words(node, prefix, results)
        results.sort(key=lambda pair: (-pair[1], pair[0]))
        return [word for word, _ in results[:k]]


if __name__ == "__main__":
    log = create_logger("trie_autocomplete_demo")

    autocomplete = TrieAutocomplete()
    words = [
        ("cat", 5), ("car", 10), ("card", 2), ("care", 7),
        ("careful", 1), ("dog", 3), ("dodge", 4),
    ]
    for word, freq in words:
        autocomplete.insert(word, freq)

    log.info("top 3 matches for 'car': %s", autocomplete.top_matches("car", k=3))
    log.info("expected ['car', 'care', 'card'] ordered by frequency descending")

    log.info("top 5 matches for 'do': %s", autocomplete.top_matches("do", k=5))
    log.info("matches for 'xyz': %s (expected empty, no such prefix)",
             autocomplete.top_matches("xyz"))
